Support beacon_klipper version >= 2.0.0, a.k.a Beacon Contact

crkochan opened the pull request:
My initial take on fixing #636.

 Beacon Contact introduced [code in their BeaconHomingHelper class](https://github.com/beacon3d/beacon_klipper/blob/730836d40f1bf02b6f1bf08649dcaacca9aee900/beacon.py#L2371-L2375) that specifically balks when `[homing_override]` is present in the Klipper config.
 
My attempt at working around this was to...

- Create a separate `beacon_contact.cfg` in the hardware probes folder
- Specify `beacon_contact` as `variable_probe_type_enabled` to be used by future g-code conditionals
- Update `machine.cfg` so that `homing_override.cfg` is not always included
- Move the `homing_override.cfg` include into `generic_probe.cfg` base macro so that every other probe can still get it

There are still a couple more changes I want to make based on the Beacon documentation that are centered on the start_print macro, but I'm opening this as a draft PR so that I can also include any feedback, since this is my first contribution.

crkochan committed:
Create beacon_contact.cfg

crkochan committed:
Rename homing_override.cfg to homing_override.cf_

crkochan committed:
Don't import homing_override by default.

crkochan committed:
Rename homing_override.cf_ to homing_override.cfg

crkochan committed:
Move homing_override include into probe configs that will use it

crkochan committed:
Don't include generic_probe.cfg

crkochan committed:
Add beacon_contact to printer.cfg template

crkochan committed:
Fix typo

crkochan committed:
Re-format heatsoak message

crkochan committed:
Add beacon max probing temp variable

…plus some whitespace clean-up

crkochan committed:
Add beacon to extruder pre-heat module

crkochan committed:
Add deactivation zhop variable for Beacon

Pretty sure this isn’t really neaded, but added just in case.

crkochan committed:
Updated beacon_contact hardware config

* Added macro include for activation and deactivation gcode
* Added macro include for homing hooks that re-implement portions of homing_override

crkochan committed:
Add Activate/Deactivate macros for beacon_contact

Macros re-purposed from vorontap because the nozzle will be touching the bed surface

crkochan committed:
Add beacon homing hooks

Pre and post homing hooks are being used to re-implement large portions of functionality from homing_override.

* Current adjustments for sensorless homing
* Endstop backoffs
* Status lighting
* Accel adjustments

crkochan committed:
Remove usused vars

crkochan committed:
Fix includes and file location

crkochan committed:
Add missing vars

crkochan committed:
Add variable declaration and missing vars

crkochan committed:
Set contact optimized startprint_actions

crkochan committed:
Move post status_leds to delay_gcode

Keeps the LEDs from quickly cyling between X and Y homing.

crkochan committed:
Forgot a variable

crkochan committed:
Fix incorrect command

crkochan committed:
Add macro variable to STATUS_LEDS

Variable can be used for conditional actions based on the active color value.

crkochan committed:
Present value as a string literal

crkochan committed:
Case and quotes adjustments

crkochan committed:
Change quote type

crkochan committed:
More quotes

crkochan committed:
Discard use of delayed_gcode

beacon_klipper has undocumented hook features that are called before or after homing either the x or y axis

crkochan committed:
Add xy hooks

crkochan committed:
Add activation macro

crkochan committed:
Add contact macro

crkochan committed:
Case change

crkochan committed:
Add recommend homing configuration

crkochan committed:
Revert status_leds

crkochan committed:
Hook improvments and minor bug fixes (#1)

* Fix incorrect command

* Add macro variable to STATUS_LEDS

Variable can be used for conditional actions based on the active color value.

* Present value as a string literal

* Case and quotes adjustments

* Change quote type

* More quotes

* Discard use of delayed_gcode

beacon_klipper has undocumented hook features that are called before or after homing either the x or y axis

* Add xy hooks

* Add activation macro

* Add contact macro

* Case change

* Add recommend homing configuration

* Revert status_leds

crkochan committed:
Add fast QGL/Z-Tilt macros

crkochan committed:
Chomp whitespace

crkochan committed:
Merge branch 'beacon-contact' into beacon-contact-testing

crkochan committed:
Add fast QGL/Z-Tilt macros (#2)

* Fix incorrect command

* Add macro variable to STATUS_LEDS

Variable can be used for conditional actions based on the active color value.

* Present value as a string literal

* Case and quotes adjustments

* Change quote type

* More quotes

* Discard use of delayed_gcode

beacon_klipper has undocumented hook features that are called before or after homing either the x or y axis

* Add xy hooks

* Add activation macro

* Add contact macro

* Case change

* Add recommend homing configuration

* Revert status_leds

* Add fast QGL/Z-Tilt macros

* Chomp whitespace

crkochan committed:
Merge branch 'beacon-contact' into beacon-contact-testing

crkochan committed:
Implement hardware check

crkochan committed:
Merge pull request #3 from crkochan/beacon-contact-testing

Add hardware check

crkochan committed:
Probing status flag and ensure beacon calibration (#4)

* Add calibration to G28 on probe activation

Guarantees some calibration occurs when home_method is set to proximity, and home_autocalibrate is set to none.

* Track probe activation status

Trying to avoid the probing hooks changing status LEDs when they shouldn’t

ksummers92 commented:
I tried your implementation and there are some default values that are not included that causes an error "Option 'home_method' is not valid in section 'beacon'". This is likely due to home_xy_position not being defined by some kind of defaults. Probably best to retrieve the zero reference position/bed center like the homing_override.cfg does and assume that as the home_xy_position.

crkochan commented:
I started working on this PR after I already had Beacon more or less up and running, so you're right, I already had `home_xy_position` defined in my overrides.cfg file.

Unfortunately, the value needs to be defined outside of a gcode macro, so I can't use the same trick that `homing_override.cfg` does.

I'm not quite sure what the optimal solution is. I can add [the documented recommended defaults](https://docs.beacon3d.com/contact/#52-configuration) into `beacon_contact.cfg`, and pick a set of coordinates that are not totally outlandish for a Voron Trident or V2, but that might not fly for users of other printers.

Maybe that, plus adding a commented beacon block to the user_templates overrides.cfg, basically some way of signaling that for contact, a user is probably going to want to have some sort of config override setup.

Just to share, this is my beacon block present in `overrides.cfg`.

```
[beacon]
serial: /dev/serial/by-id/usb-Beacon_Beacon_RevH_9AA5D66B5154354D38202020FF0A0B36-if00
x_offset: 0
y_offset: 26.5
mesh_main_direction: x
mesh_runs: 2
accel_axes_map: x, y, z
home_xy_position: 150, 150
home_z_hop: 5
home_z_hop_speed: 30
home_xy_move_speed: 300
autocal_speed: 5
home_method: proximity
home_method_when_homed: proximity
home_autocalibrate: never
```

crkochan committed:
Add more quick start defaults from Beacon docs

If home_xy_position is unset, Klipper will throw an `Option 'home_method' is not valid` error message.

Adding it, plus other defaults from the Beacon Contact docs.

crkochan committed:
Beacon contact testing (#5)

* Add calibration to G28 on probe activation

Guarantees some calibration occurs when home_method is set to proximity, and home_autocalibrate is set to none.

* Track probe activation status

Trying to avoid the probing hooks changing status LEDs when they shouldn’t

* Calibrate after tilt

* Update model prior to tilt

* Fix inconsistent indents

zanmoskotevc commented:
Hey @crkochan, any update on this? I'd love to see support for this added.

crkochan commented:
None from my side. I've been running a series of post-kit-build prints for some new parts and future upgrades using my fork branch, and the pieces have been turning out just fine on my V2.4.

It may not be up to snuff in terms of being merged into mainline because of the need for users to add Contact-specific settings that are perhaps unique to their machine, or class of printer, into their overrides file in order to avoid the potential for cryptic and/or misleading error messages on Klipper start that would come from the Beacon module.

I'm happy to implement any feedback from the maintainers.

KazachThi commented:
Hey, @crkochan I have been using your setup to run beacon contact with no problems. All is working as intended. 
One thing (and i might be wrong on this one), we have one unnecessary beacon contact calibration.
Order in my case :
 - Homing
 - pre-heating hotend
 - bed heating
 - chamber heating
 - Beacon contact calibration (first)
 - GQL
 - Beacon contact calibration (second)
 - Beacon contact calibration (third)
 - Bed mesh
 - Beacon contact calibration (fourth)
 - Heating hotend
 - Nozzle clean
 - Prime line
 - Printing

So, second and third are redundant ?
Hope this helps.

crkochan commented:
> So, second and third are redundant ? Hope this helps.

Yes. Per the Beacon docs, QGL/z tilt adjust and bed mesh should be book-ended by calibrations. 

I'm not sure where the extra calibrations are coming from in your case.

crkochan committed:
Merge branch 'main' into beacon-contact

laukejas commented:
Has there been any progress with this?

Frix-x commented:
Hello, thanks for the PR. I'm back on the subject and bought a beacon to test it. I'll merge it as soon as I get it installed :)

Frix-x commented on config/hardware/probes/beacon_contact.cfg:
```diff
@@ -0,0 +1,59 @@
+contact_activate_gcode: _CONTACT_MACRO STAGE=activate
+contact_deactivate_gcode: _CONTACT_MACRO STAGE=deactivate
+
+[gcode_macro _beacon_contact_hardware_check]
+gcode:
+    {% if 'mcu beacon' in printer %}
+        {% set version = printer['mcu beacon'].mcu_version.split(' ')|last %}
+        {% set major = version.split('.')|first|int %}
+        {% if version < '2' %}
+            {action_raise_error('Beacon firmware is too old. Found: %s, Expected: 2.0.0 or higher' % version)}
+        {% endif %}
+    {% else %}
+        {action_raise_error('Beacon MCU was not found')}
+    {% endif %}
+
+[delayed_gcode _do_beacon_contact_hardware_check]
+initial_duration: 5
+gcode:
+    _beacon_contact_hardware_check
```
Is it possible to move this check in the startup.cfg file? There's already a few things that are checked at Klippain startup and this could be added in the _INIT_CHECKPROBECONF macro that is done exactly for this purpose.

https://github.com/Frix-x/klippain/blob/main/macros/miscs/startup.cfg

Frix-x commented on macros/base/probing/beacon_probe.cfg:
```diff
@@ -0,0 +1,62 @@
+        { action_respond_info('Extruder temperature target of %.1fC is too high for Beacon probing, lowering to %.1fC' % (TARGET_TEMP, beacon_max_probing_temp)) }
+        M106 S255 ; 100% the part cooling fan to help the extruder cooling
+        M109 S{beacon_max_probing_temp}
+        M106 S0   ; Stop the part cooling fan
+    {% else %}
+        # Temperature target is already low enough, but nozzle may still be too hot
+        {% if ACTUAL_TEMP > beacon_max_probing_temp + 3 %}
+            M106 S255 ; 100% the part cooling fan to help the extruder cooling
+            TEMPERATURE_WAIT SENSOR=extruder MAXIMUM={beacon_max_probing_temp}
+            M106 S0   ; Stop the part cooling fan
+        {% endif %}
+    {% endif %}
+
+    G28 Z METHOD=CONTACT CALIBRATE=1
+
+
+[gcode_macro DEACTIVATE_PROBE]
+description: Revert the machine to a normal state after probing
+gcode:
+    {% set beacon_deactivation_zhop = printer["gcode_macro _USER_VARIABLES"].beacon_deactivation_zhop %}
+    {% set Sz = printer["gcode_macro _USER_VARIABLES"].z_drop_speed * 60 %}
+
+    G28 Z METHOD=CONTACT CALIBRATE=1
+
+    # Check and restore the nozzle temperature if needed
+    # Small Z hop to avoid restoring the temperature directly on the PEI
+    {% set z_safe = printer.toolhead.position.z + beacon_deactivation_zhop %}
+    {% if z_safe > printer.toolhead.axis_maximum.z %}
+        {% set z_safe = printer.toolhead.axis_maximum.z %}
+    {% endif %}
+    G90
+    G1 Z{z_safe} F{Sz}
+
+    # Then restoring the temperature
+    {% set old_target_temperature = printer["gcode_macro ACTIVATE_PROBE"].temperature %}
+    M109 S{old_target_temperature}
+
+    RESTORE_GCODE_STATE NAME=BEFORE_BEACON_ACTION
+
+    SET_GCODE_VARIABLE MACRO=ACTIVATE_PROBE VARIABLE=is_active VALUE=False
```
I think this duplication of ACTIVATE/DEACTIVATE_PROBE doesn't make sense and could be avoided. In fact, these original macros were designed to be generic and accept any type of probe.

If I understand correctly, you made this duplication choice because you added `[include ../homing/homing_override.cfg]` to the generic_probe.cfg file. But maybe you can move that line to any probe type that needs the default homing_override sequence, and keep the generic_probe.cfg really generic to accept that content for beacon as well.

See my next comments.

Frix-x commented on macros/base/probing/generic_probe.cfg:
```diff
@@ -4,6 +4,8 @@
 #  - Dockable probe: need to be attached/docked when used
 #  - ... can be improved depending of new probes needs ...
 
+[include ../homing/homing_override.cfg]
```
Move this line into each probe type that require it.

Frix-x commented on macros/base/probing/generic_probe.cfg:
```diff
@@ -4,6 +4,8 @@
 
+[include ../homing/homing_override.cfg]
+
 [gcode_macro ACTIVATE_PROBE]
```
And add the beacon specificity in these macros.

Frix-x commented on macros/base/probing/overrides/beacon_probe_overrides.cfg:
```diff
@@ -0,0 +1,79 @@
+# Improvements are from the ANNEX Discord and allow for QGL/Tilt in scan mode
+#############################################################################
+
+[gcode_macro FAST_QUAD_GANTRY_LEVEL]
```
I'm not a fan of using a new `FAST_...` macro for a duplication of the same feature that QGL already provide but adapted to be faster. I'm pretty sure everyone will always want to use the fast version.

So why not merging this in the original overrides files with a check on the `beacon_contact` (and also standard `beacon`?) probe type to do the fast version?
`https://github.com/Frix-x/klippain/blob/main/macros/base/probing/overrides/qgl.cfg`
`https://github.com/Frix-x/klippain/blob/main/macros/base/probing/overrides/z_tilt.cfg`

This way, when running the standard QGL/Z_TILT macro that is already available in the frontends, it will always work for everyone, using the fast way if you can go fast with a beacon.

Frix-x commented on macros/base/probing/overrides/beacon_probe_overrides.cfg:
```diff
@@ -0,0 +1,79 @@
+    {% endif %}
+
+
+[gcode_macro FAST_Z_TILT_ADJUST]
```
Same comment here about the `FAST_...` version of the Z_TILT_ADJUST macro

Frix-x reviewed (changes requested):
So I've made a review of the code (I haven't tested it yet as I'm still waiting for my beacon to arrive), but I already have some changes requested to simplify the code a bit.

I really want to keep it simple for Klippain users to use it and avoid duplicating macros, etc...

tavalin commented:
Hi, is there any progress on this or plans to progress this?  The Beacon contact functionality has been out for more than 6 months now so it's a shame that Klippain doesn't yet officially support it

sima-it commented:
So I assume, this will not be intregrated then. Is the Poject dead?

gueee commented:
> So I assume, this will not be intregrated then. Is the Poject dead?

it seems to be dead at that point.
nevertheless, I forked the repo and will extend the project, but I can't say when I'm ready to publish it.

for now I pulled my own repo, so it's not marked dirty and beacon contact and AWD are working for me.

sima-it commented:
thx mate, 
i'm running also the changes from [crkochan] with no problems so far just the dirty repo keeps bothering me ;)

gueee commented:
I already extended klippain a bit, especially regarding bed sizes with a dedicated configuration option to choose any bed size, which calculates the offsets for bed mesh and the configuration option for multiple motors on x and y, including tmc driver settings for the additional steppers.

I'm also trying to create a web based gui for automatically generating the configs, or rather the overrides. 

if you're a little patient I'm gonna publish it, probably under a new name, since there are too many changes that somehow left the path of klippain...

sima-it commented:
looking forward to test it out, thanks for your effort

neptunus0001 commented:
@Frix-x Can you give klippain users an indication of when this feature will be integrated? It would be great to get some feedback from you.

ormarek commented:
Any news on that? It's been over a year
@crkochan @gueee @Frix-x

gueee commented:
> Any news on that? It's been over a year
> @crkochan @gueee @Frix-x

I guess Frix-x hasn't time to maintain the project anymore, but I started a similar project, maybe you wanna have a look at it?!

https://github.com/gueee/gschpoozi

a warning, it's in alpha state, but I guess I can get it to a stable beta today!

it's mostly functional already though.

ormarek commented:
@gueee hey, that looks super promising! Unfortunately I have only one printer and currently no budget for fixes and time to play with stuff so I might have to wait a little until I have some spare cash for a printer before trying that out (reason is alpha/beta).

But I started your repository! Interactive setup sounds like very neat way to set up things

mjonuschat commented:
Closed by #721
